Add tf.io.browserIndexedDB

caisq opened the pull request:
for saving model artifacts to browser IndexedDB and loading them.

Towards: https://github.com/tensorflow/tfjs/issues/13

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-core/1017)
<!-- Reviewable:end -->

arunikayadav42 commented:
Hello all I am e beginner in open source and have a basic knowledge of javascript and want to get started with contributing to this project!Can somebody please help me?I also know a bit of NLP and deep learning.

caisq commented:
@arunikayadav42 Great to see you are interested in contributing. Maybe you can take a look at the list of our issues with the "help wanted" tag and see which ones interest you. https://github.com/tensorflow/tfjs/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22

nsthorat commented:
Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---

*[src/io/indexed_db.ts, line 35 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgb2_g7POn5uECMjrd:-LBgb2_g7POn5uECMjre:b-3zllce) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L35)):*
> ```XML
>   return new Promise<void>((resolve, reject) => {
>     const deleteRequest = idbFactory.deleteDatabase(DATABASE_NAME);
>     deleteRequest.onsuccess = () => {
> ```

simplify these:

`deleteRequest.onsuccess = () => resolve();`

---

*[src/io/indexed_db.ts, line 45 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgbASWCzjXps8_Phbp:-LBgbASWCzjXps8_Phbq:b-ad5zby) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L45)):*
> ```XML
> 
> function getIndexedDBFactory(): IDBFactory {
>   if (typeof window === 'undefined') {
> ```

Can you add an environment flag for the environment? Maybe called "IS_NODE" so we don't have these checks sprinkled everywhere?

Add in 3 places:
https://github.com/tensorflow/tfjs-core/blob/master/src/environment.ts#L30
https://github.com/tensorflow/tfjs-core/blob/master/src/environment.ts#L53
https://github.com/tensorflow/tfjs-core/blob/master/src/environment.ts#L324

---

*[src/io/indexed_db.ts, line 55 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgc-1v8DX-eguq34co:-LBgc-1v8DX-eguq34cp:b-yj4xo4) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L55)):*
> ```XML
>   // tslint:disable-next-line:no-any
>   const theWindow: any = window;
>   return theWindow.indexedDB || theWindow.mozIndexedDB ||
> ```

not all browsers support this, should we throw if it's null?

---

*[src/io/indexed_db.ts, line 83 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgd0fs0CDSzk6wbSyQ:-LBgd0fs0CDSzk6wbSyR:byuak4p) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L83)):*
> ```XML
> 
>     return new Promise<SaveResult>((resolve, reject) => {
>       const modelArtifactsInfo: ModelArtifactsInfo = {
> ```

this seems to be pretty common -- maybe pull out into a util.

---

*[src/io/indexed_db.ts, line 96 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgc5G82iOTh6dShL4w:-LBgc5G82iOTh6dShL4x:bcaem50) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L96)):*
> ```XML
> 
>       // Create the schema.
>       openRequest.onupgradeneeded = () => {
> ```

`openRequest.onupgradeneeded = () => this.setUpDatabase(openRequest);`

---

*[src/io/indexed_db.ts, line 113 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcl2f7phaQ4omcZdp:-LBgcl2f7phaQ4omcZdq:b-4pzl5x) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L113)):*
> ```XML
>           reject(putRequest.error);
>         };
>         tx.oncomplete = () => {
> ```

`tx.oncomplete = () => db.close();` 

follow this convention elsewhere where you can as well, you dont need the {} for single-line body callbacks.

---

*[src/io/indexed_db.ts, line 127 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcEyH0q-V7vGDpbMk:-LBgcEyH0q-V7vGDpbMl:bcaem50) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L127)):*
> ```XML
>       const openRequest = this.indexedDB.open(DATABASE_NAME, DATABASE_VERSION);
> 
>       openRequest.onupgradeneeded = () => {
> ```

`openRequest.onupgradeneeded = () => this.setUpDatabase(openRequest);`

---

*[src/io/indexed_db.ts, line 138 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcIKNDJGKWbuJY2hI:-LBgcIKNDJGKWbuJY2hJ:br3rpku) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L138)):*
> ```XML
>         const getRequest = store.get(this.modelPath);
>         getRequest.onsuccess = () => {
>           if (getRequest.result === undefined) {
> ```

double equals null

---

*[src/io/indexed_db.ts, line 149 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcK6Y4VT-biUy-eTw:-LBgcK6Y4VT-biUy-eTx:b38t562) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L149)):*
> ```XML
>           reject(getRequest.error);
>         };
>         tx.oncomplete = () => {
> ```

`tx.oncomplete = () => db.close();`

---

*[src/io/indexed_db.ts, line 153 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcNIz6DUD_0ThvGA3:-LBgcNJ-3PtMzEbqmxsy:bsmzs4v) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L153)):*
> ```XML
>         };
>       };
>       openRequest.onerror = (error) => {
> ```

`openRequest.onerror = error => reject(openRequest.error);`

---

*[src/io/indexed_db.ts, line 166 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcdN05I93zE54uPzt:-LBgcdN05I93zE54uPzu:bbvte2) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L166)):*
> ```XML
> 
> /**
>  * Factory function for browser IndexedDB IOHandler.
> ```

Same comment here: "Creates a browser IndexedDB IOHandler for saving models."

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017)*
<!-- Sent from Reviewable.io -->

bileschi commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/io/indexed_db.ts, line 59 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgnEuhBGCVMh0FQeLP:-LBgnEuhBGCVMh0FQeLQ:b-wl9fa5) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L59)):*
> ```XML
>       theWindow.shimIndexedDB;
> }
> ```

Please add something analogous to the local_storage comment here.
/**
 * IOHandler subclass: Browser IndexedDB.
 *
 * See the doc string to `browserIndexedDB` for more details.
 */

---

*[src/io/indexed_db.ts, line 161 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgqJ5b45m_p71RVWOY:-LBgqJ5b45m_p71RVWOZ:b-mdc8me) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L161)):*
> ```
>  db.createObjectStore(OBJECT_STORE_NAME, {keyPath: 'modelPath'});
> ```
This method can fail.  It seems it's ok if it fails because the objectStore already exists, but some of the other reasons should maybe be caught?
https://developer.mozilla.org/en-US/docs/Web/API/IDBDatabase/createObjectStores

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: 0 of 6 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/io/indexed_db.ts, line 35 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgb2_g7POn5uECMjrd:-LBnWgRf1jOZOAg20XY1:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L35)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

simplify these:

`deleteRequest.onsuccess = () => resolve();`
</blockquote></details>

Done.

---

*[src/io/indexed_db.ts, line 45 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgbASWCzjXps8_Phbp:-LBq2L_F8zAc7oMXUZIu:bdqxnc0) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L45)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Can you add an environment flag for the environment? Maybe called "IS_NODE" so we don't have these checks sprinkled everywhere?

Add in 3 places:
https://github.com/tensorflow/tfjs-core/blob/master/src/environment.ts#L30
https://github.com/tensorflow/tfjs-core/blob/master/src/environment.ts#L53
https://github.com/tensorflow/tfjs-core/blob/master/src/environment.ts#L324
</blockquote></details>

I don't feel I'm familiar with `environment.ts` enough to be able to do it correctly. But I agree we should  deduplicate and centralize these checks. I added a TODO item to get that done in a later PR, maybe with your help. Let me know whether that's okay.

---

*[src/io/indexed_db.ts, line 55 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgc-1v8DX-eguq34co:-LBnVAPt0EaRP2PB6J56:b-qu6n94) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L55)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

not all browsers support this, should we throw if it's null?
</blockquote></details>

Good point. Done.

---

*[src/io/indexed_db.ts, line 59 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgnEuhBGCVMh0FQeLP:-LBnVWyI7m-uRXDIn5OC:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L59)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

Please add something analogous to the local_storage comment here.
/**
 * IOHandler subclass: Browser IndexedDB.
 *
 * See the doc string to `browserIndexedDB` for more details.
 */
</blockquote></details>

Done.

---

*[src/io/indexed_db.ts, line 83 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgd0fs0CDSzk6wbSyQ:-LBpyZkc5F4MtFm9vFlP:blup33p) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L83)):*
> common
Good point. Done.

---

*[src/io/indexed_db.ts, line 96 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgc5G82iOTh6dShL4w:-LBq-Rul893IBcsqq1Ai:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L96)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

`openRequest.onupgradeneeded = () => this.setUpDatabase(openRequest);`
</blockquote></details>

Done.

---

*[src/io/indexed_db.ts, line 113 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcl2f7phaQ4omcZdp:-LBq-o7NBc4T7Wc24Nsb:bupftgs) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L113)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

`tx.oncomplete = () => db.close();` 

follow this convention elsewhere where you can as well, you dont need the {} for single-line body callbacks.
</blockquote></details>

Done. Simplified all places like this in this file.

---

*[src/io/indexed_db.ts, line 127 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcEyH0q-V7vGDpbMk:-LBq-rAz3IkrZQd21GjG:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L127)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

`openRequest.onupgradeneeded = () => this.setUpDatabase(openRequest);`
</blockquote></details>

Done.

---

*[src/io/indexed_db.ts, line 138 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcIKNDJGKWbuJY2hI:-LBq-tl-9cByyzuAExH_:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L138)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

double equals null
</blockquote></details>

Done.

---

*[src/io/indexed_db.ts, line 149 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcK6Y4VT-biUy-eTw:-LBq-uD2CaMyXIodBZL3:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L149)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

`tx.oncomplete = () => db.close();`
</blockquote></details>

Done.

---

*[src/io/indexed_db.ts, line 153 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcNIz6DUD_0ThvGA3:-LBq-uf2Dm8xxTdZV34L:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L153)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

`openRequest.onerror = error => reject(openRequest.error);`
</blockquote></details>

Done.

---

*[src/io/indexed_db.ts, line 161 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgqJ5b45m_p71RVWOY:-LBq2zxRDYlEa-e3Yisc:bvtx7e6) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L161)):*
> fail

If this method fails, it should throw an error and the error will surface in the BrowserIndexedDB.save and .load calls.

---

*[src/io/indexed_db.ts, line 166 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcdN05I93zE54uPzt:-LBq01R36EAl7LwF9xk5:bnloaaz) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L166)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Same comment here: "Creates a browser IndexedDB IOHandler for saving models."
</blockquote></details>

Done. Added the word "loading" as the returned IOHandler also supports loading.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017)*
<!-- Sent from Reviewable.io -->

pyu10055 commented:
Review status: 0 of 6 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/io/indexed_db.ts, line 153 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcNIz6DUD_0ThvGA3:-LC1XAfP4KgBVgcDO4m5:b3s6rmr) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L153)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

still has braces

---

*[src/io/indexed_db.ts, line 75 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1VpAu9EE6KDVqOMuc:-LC1VpAu9EE6KDVqOMud:brd9jqo) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L75)):*
> ```
> !modelPath
> ```
!modelPath also check modelPath === null

---

*[src/io/indexed_db.ts, line 94 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1WLza3_S2qiJJ5HRv:-LC1WLza3_S2qiJJ5HRw:by2mb4a) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L94)):*
> ```XML
>           getModelArtifactsInfoForKerasJSON(modelArtifacts);
> 
>       const openRequest = this.indexedDB.open(DATABASE_NAME, DATABASE_VERSION);
> ```

the database code for save and load method are quite similar, maybe better to refactor to a wrapper class that
takes in transaction type and onsuccess function.

---

*[src/io/indexed_db.ts, line 122 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1WhamAoDHgOzIcjIe:-LC1Whan8CQTLyLtNEoA:b-dlbhf6) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L122)):*
> ```
> readwrite
> ```
will read be enough?

---

*[src/io/indexed_db.ts, line 127 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1Wtp58ayE4yI6SNuT:-LC1Wtp58ayE4yI6SNuU:buirxu3) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L127)):*
> ```XML
>         const getRequest = store.get(this.modelPath);
>         getRequest.onsuccess = () => {
>           if (getRequest.result == null) {
> ```

use ===

---

*[src/io/indexed_db_test.ts, line 135 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1XSTe7ZKibdUOKz8l:-LC1XSTe7ZKibdUOKz8m:b-6wwpuj) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db_test.ts#L135)):*
> ```
> async
> ```
since you are using done, is the async still needed? 

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: 0 of 6 files reviewed at latest revision, 18 unresolved discussions.

---

*[src/io/indexed_db.ts, line 153 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcNIz6DUD_0ThvGA3:-LC1kcbIARALwXrKBQrn:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L153)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

still has braces
</blockquote></details>

Done.

---

*[src/io/indexed_db.ts, line 94 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1WLza3_S2qiJJ5HRv:-LC1pbiM-3vAN4oLboB1:bt4ox1i) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L94)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

the database code for save and load method are quite similar, maybe better to refactor to a wrapper class that
takes in transaction type and onsuccess function.
</blockquote></details>

Done. Thanks for the suggestion.

---

*[src/io/indexed_db.ts, line 127 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1Wtp58ayE4yI6SNuT:-LC1lFL22WJtLnCOw85Z:b-7vulrk) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L127)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

use ===
</blockquote></details>

Done. Using `=== undefined`.

---

*[src/io/indexed_db_test.ts, line 135 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1XSTe7ZKibdUOKz8l:-LC1l6_m-PjsXMn0NW-y:bnmx8hm) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db_test.ts#L135)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

> ```
> async
> ```
since you are using done, is the async still needed? 
</blockquote></details>

Good point. `async` is no longer needed in these tests. Removed them.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017)*
<!-- Sent from Reviewable.io -->

davidsoergel commented:
Reviewed 4 of 5 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[src/io/indexed_db.ts, line 29 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4ZXIh5cji2rWhjYKx:-LC4ZUy0Cx-J3upuH8oR:b-fej8b5) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2fd54f08311e6c8aa59b555b19bc8f72a835bf76/src/io/indexed_db.ts#L29)):*
> ```
> tensorfowjs
> ```
typo

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: 5 of 6 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[src/io/indexed_db.ts, line 29 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4ZXIh5cji2rWhjYKx:-LC4bal9BLEaZxbwgpRH:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2fd54f08311e6c8aa59b555b19bc8f72a835bf76/src/io/indexed_db.ts#L29)):*
<details><summary><i>Previously, davidsoergel (David Soergel) wrote…</i></summary><blockquote>

> ```
> tensorfowjs
> ```
typo
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Nice work! Sorry for delay - things are a little busy. Left a few comments but nothing blocking, thus LGTM

---

Review status: 5 of 6 files reviewed at latest revision, 19 unresolved discussions.

---

*[src/io/browser_files.ts, line 39 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4ae851I8xLVs2uGKQ:-LC4ae851I8xLVs2uGKR:bmmy82c) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2fd54f08311e6c8aa59b555b19bc8f72a835bf76/src/io/browser_files.ts#L39)):*
> ```XML
> 
>   constructor(fileNamePrefix?: string) {
>     // TODO(cais): Use central environment flag when it's available.
> ```

todo's are easy to forget about. better to file an issue and link to it: e.g. TODO(https://github.com/tensorflow/tfjs/issues/999): .../

---

*[src/io/indexed_db.ts, line 59 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgnEuhBGCVMh0FQeLP:-LC4eeKv3u1B7Cmdx8uf:b-x7glde) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L59)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Since we are not going to expose this class, I think there is no need for the jsdoc, because it's another thing to keep in sync if we rename things.

Also `BrowserIndexedDB implements IOHandler` is self-documenting and the jsdoc just repeats that.

---

*[src/io/indexed_db.ts, line 21 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4dSEw-EMt5QKzAp0i:-LC4dSEw-EMt5QKzAp0j:b-2gt0nn) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db.ts#L21)):*
> ```XML
> import {getModelArtifactsInfoForKerasJSON} from './io_utils';
> import {IOHandler, ModelArtifacts, ModelArtifactsInfo, SaveResult} from './types';
> ```

remove extra newline

---

*[src/io/indexed_db.ts, line 44 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4dq0dDgO1hdpjUjqR:-LC4dq0dDgO1hdpjUjqS:b-548552) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db.ts#L44)):*
> ```XML
>   // TODO(cais): Use central environment flag when it's available.
>   if (typeof window === 'undefined') {
>     // TODO(cais): Add more info about what IOHandler subtypes are available.
> ```

remove todo and file an issue. this way we open the door to discussion and potential contributors that can tackle this problem.

---

*[src/io/indexed_db.ts, line 120 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4chtX84YKrIvQgwcn:-LC4chtX84YKrIvQgwco:b2jr79q) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db.ts#L120)):*
> ```XML
> 
>       openRequest.onsuccess = () => {
>         const db = openRequest.result as IDBDatabase;
> ```

double checking `IDBDatabase` and `IDBFactory` are built-in types right?

---

*[src/io/indexed_db.ts, line 178 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4fpbN8rcjjo7lACVf:-LC4fpbN8rcjjo7lACVg:b-z7yi5i) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db.ts#L178)):*
> ```XML
>  *   which can be used with, e.g., `tf.Model.save`.
>  */
> export function browserIndexedDB(modelPath: string): BrowserIndexedDB {
> ```

since there are no special methods on specific `IOHandlers` (by design), let's return `IOHandler` here instead of `BrowserIndexedDB` and update @returns to just say it returns an IOHandler that stores data in IndexedDB and it should be used with tf.Model.save.

---

*[src/io/indexed_db_test.ts, line 25 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4gPLt6YQUBllXRXlv:-LC4gPLt6YQUBllXRXlw:b-hpkpoy) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db_test.ts#L25)):*
> ```XML
> import {deleteDatabase} from './indexed_db';
> 
> describe('IndexedDB', () => {
> ```

describeWithFlags('IndexedDB', CPU_ENVS, () => { ...} to constraint it to run only on cpu, otherwise this test will fail on node environment, Also no need to run this explicitly on webgl either. We will need to fix the other tests (e.g. localStorage) to also run only on cpu. See https://github.com/tensorflow/tfjs/issues/279

---

*[src/io/local_storage.ts, line 38 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4bt5X2QSN5voQZ0GI:-LC4bt5X2QSN5voQZ0GJ:bgdyzdh) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/local_storage.ts#L38)):*
> ```XML
> export function purgeLocalStorageArtifacts(): string[] {
>   // TODO(cais): Use central environment flag when it's available.
>   if (!(window && window.localStorage)) {
> ```

typeof window === 'undefined'

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-:-LC4gxgHEh8QLh7TcDeN:b-2u8pi6)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: 5 of 6 files reviewed at latest revision, 25 unresolved discussions, all commit checks successful.

---

*[src/io/indexed_db.ts, line 45 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgbASWCzjXps8_Phbp:-LC4iOxhAk81rg4SBtji:brxp9sd) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L45)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

I don't feel I'm familiar with `environment.ts` enough to be able to do it correctly. But I agree we should  deduplicate and centralize these checks. I added a TODO item to get that done in a later PR, maybe with your help. Let me know whether that's okay.
</blockquote></details>

Sounds good!

---

*[src/io/indexed_db.ts, line 138 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgcIKNDJGKWbuJY2hI:-LC4id_319vFjmqqLaiI:b19p7bb) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L138)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I still see triple equals undefined. Prefer double equals null.

---

*[src/io/indexed_db.ts, line 127 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1Wtp58ayE4yI6SNuT:-LC4ijG1Cz1-icDOq5Ie:bw08lji) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L127)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Done. Using `=== undefined`.
</blockquote></details>

You shouldn't be using === undefined here (see comment above).

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-:-LC4iuUZ0UaYmJ2NB1CM:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: 5 of 6 files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[src/io/indexed_db.ts, line 59 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LBgnEuhBGCVMh0FQeLP:-LC4lL6H48qEHepKBdw9:b6lffqc) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a2d5950c9e4b0f577f33db09cc3b9a4ecf0ecc41/src/io/indexed_db.ts#L59)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Since we are not going to expose this class, I think there is no need for the jsdoc, because it's another thing to keep in sync if we rename things.

Also `BrowserIndexedDB implements IOHandler` is self-documenting and the jsdoc just repeats that.
</blockquote></details>

It doesn't hurt to have the doc string here. 

---

*[src/io/indexed_db.ts, line 75 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1VpAu9EE6KDVqOMuc:-LC4oHlT36KBQURoDNpE:bhm4t3v) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L75)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

> ```
> !modelPath
> ```
!modelPath also check modelPath === null
</blockquote></details>

=== null is covered in == null

---

*[src/io/indexed_db.ts, line 122 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1WhamAoDHgOzIcjIe:-LC4ls748rz3nafHyJim:b-x9d7ju) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L122)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

> ```
> readwrite
> ```
will read be enough?
</blockquote></details>

Using more restrictive permissions in the new helper function (`databaseAction`).

---

*[src/io/indexed_db.ts, line 127 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1Wtp58ayE4yI6SNuT:-LC4mLYS0T3-8RnU4FpJ:b-k2dae7) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L127)):*
> === undefined

OK. Sorry about the back and forth. I agree with Nikhil that `== null` is better here, because in the lines below we are going to access the `modelTopologhy field of `getRequest.result`.

---

*[src/io/indexed_db.ts, line 21 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4dSEw-EMt5QKzAp0i:-LC4lQ-J5IHL_6S5neHR:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db.ts#L21)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

remove extra newline
</blockquote></details>

Done.

---

*[src/io/indexed_db.ts, line 44 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4dq0dDgO1hdpjUjqR:-LC4lff13sizyc7tusJ9:b-eetnf1) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db.ts#L44)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

remove todo and file an issue. this way we open the door to discussion and potential contributors that can tackle this problem.
</blockquote></details>

Filed. https://github.com/tensorflow/tfjs/issues/282 

I'll keep the TODO item here just so it's easy to understand what the issues is talking about.

---

*[src/io/indexed_db.ts, line 120 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4chtX84YKrIvQgwcn:-LC4lpRP6ZRyLFo0zMvc:b-pu83fa) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db.ts#L120)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

double checking `IDBDatabase` and `IDBFactory` are built-in types right?
</blockquote></details>

They seem to be built-in types for TypeScript.

---

*[src/io/indexed_db.ts, line 178 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4fpbN8rcjjo7lACVf:-LC4moKEEIIn65D0y5dF:bjsnxkd) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db.ts#L178)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

since there are no special methods on specific `IOHandlers` (by design), let's return `IOHandler` here instead of `BrowserIndexedDB` and update @returns to just say it returns an IOHandler that stores data in IndexedDB and it should be used with tf.Model.save.
</blockquote></details>

I agree. Done. Using the same pattern elsewhere as well.

---

*[src/io/indexed_db_test.ts, line 25 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4gPLt6YQUBllXRXlv:-LC4nlbv7-Kcu3AXDtMk:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/indexed_db_test.ts#L25)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

describeWithFlags('IndexedDB', CPU_ENVS, () => { ...} to constraint it to run only on cpu, otherwise this test will fail on node environment, Also no need to run this explicitly on webgl either. We will need to fix the other tests (e.g. localStorage) to also run only on cpu. See https://github.com/tensorflow/tfjs/issues/279
</blockquote></details>

Done.

---

*[src/io/local_storage.ts, line 38 at r5](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC4bt5X2QSN5voQZ0GI:-LC4mDJo7Q473PiVVaZ4:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a8e1e7cfc8613727d46e16635ea0f549391f74db/src/io/local_storage.ts#L38)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

typeof window === 'undefined'
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017)*
<!-- Sent from Reviewable.io -->

pyu10055 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 2 of 5 files at r3, 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 25 unresolved discussions, all commit checks successful.

---

*[src/io/indexed_db.ts, line 75 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-LC1VpAu9EE6KDVqOMuc:-LC58ZEK7ahQ4TBLun_k:b-7xezaq) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2055ee731b5f4718cbcc1c8c37753fe34022c054/src/io/indexed_db.ts#L75)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

=== null is covered in == null
</blockquote></details>

according ts style guide, always use === http://go/tsstyle#equality-checks

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1017#-:-LC59CpMEf4Eo-E45Dvt:b-jbsf8z)*
<!-- Sent from Reviewable.io -->
